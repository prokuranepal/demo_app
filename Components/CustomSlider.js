import React, { useState } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import Slider from 'react-native-custom-slider';

const CustomSlider = props => {

  const defaultValue = 0
  const maximumValue = 3;
  const length = 4
  const sliderRatio = maximumValue / (length - 1);
  const [state, setState] = useState({
    value: sliderRatio * defaultValue,
    adjustSign: 1
  });
  const sliderOptions = [
    { value: 0, label: 'Eat' },
    { value: 1, label: 'Don\'t Prefer' },
    { value: 2, label: 'Intolerant' },
    { value: 3, label: 'Severe{\n}Allergy' },
  ];

  const choice = (v) => {
    const halfRatio = sliderRatio / 2;
    // console.log(halfRatio);
    let i = 0;
    for (; ;) {
      if ((v < sliderRatio) || (v <= 0)) {
        if (v >= halfRatio) {
          i++;
        }
        break;
      }
      v = v - sliderRatio;
      i++;
    }
    let value = sliderRatio * i;

    //Move the slider
    value = value + (state.adjustSign * 0.000001);//enforce UI update
    let sign = 1;
    if (state.adjustSign > 0) {
      sign = -1
    } else {
      sign = 1
    }
    // console.log(value);
    setState({ ...state, value: value, adjustSign: sign })
    props.selectState(Math.round(value))
  }


  return (
    <View>
      <View style={styles.sliderContainer}>

        <Slider
          value={state.value}
          minimumValue={0}
          maximumValue={3}
          onSlidingComplete={(value) => choice(value)}
          trackStyle={styles.track}
          thumbStyle={styles.thumb}
          minimumTrackTintColor="#464F7A"
          maximumTrackTintColor="#464F7A"
        />

      </View>
      <View style={styles.itemWrapper}>
        {/* {sliderOptions.map((option) => {
          <Text>{option.label}</Text>
        })} */}
        <Text style={styles.item}>Eat</Text>
        <Text style={styles.item}>Don't Prefer</Text>
        <Text style={styles.item}>Intolerant</Text>
        <Text style={styles.item}>Severe{"\n"}Allergy</Text>
      </View>
    </View>
  )
}

const styles = StyleSheet.create({
  itemWrapper: {
    marginLeft: 10,
    justifyContent: 'space-between',
    alignSelf: 'stretch',
    flexDirection: 'row',
  },
  item: {
    color: '#464F7A'
  },
  sliderContainer: {
    marginTop: 20,
    borderLeftWidth: 10,
    borderRightWidth: 15,
    height: 40,
    paddingBottom: 20,
    borderRadius: 20,
    backgroundColor: '#464F7A',
    borderColor: '#464F7A',
  },
  track: {
    
  },
  thumb: {
    width: 20,
    height: 20,
    backgroundColor: '#464F7A',
    borderColor: 'white',
    borderWidth: 5,
    borderRadius: 10,
  }
});

export default CustomSlider;