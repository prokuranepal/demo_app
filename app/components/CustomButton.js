import React from 'react';
import { View, TouchableOpacity, Text, StyleSheet } from 'react-native';

import PropTypes from 'prop-types';

import {fontSize} from '../theme/fonts';
import {colors} from '../theme/themeColors';

const CustomButton = (props) => (
  <View style={props.container_style}>
    <TouchableOpacity
      style={[styles.saveStyle, props.buttonStyle]}
      onPress={props.pressHandler}
      data-test="touchableComp">
      <Text style={{ ...styles.text,color: props.color}} data-test="textData">{props.title}</Text>
    </TouchableOpacity>
  </View>
)

const styles = StyleSheet.create({
  text: {
    fontSize: fontSize.buttonText
  },
  saveStyle: {
    borderRadius: 40,
    paddingHorizontal: 60,
    paddingVertical: 10,
    backgroundColor: colors.primaryButton
  }
})

CustomButton.propTypes = {
  buttonStyle: PropTypes.object,
  containerStyle: PropTypes.object,
  color: PropTypes.string,
  title: PropTypes.string,
  pressHandler: PropTypes.func
}


export default CustomButton;