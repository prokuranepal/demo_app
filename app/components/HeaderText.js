import React from "react";
import { Text, StyleSheet } from "react-native";

import PropTypes from 'prop-types';

import {fontSize} from '../theme/fonts';
import {colors} from '../theme/themeColors'

const HeaderText = props => (
  <Text {...props} style={[styles.textHeading, props.style]} data-test="headerText">
    {props.children}
  </Text>
);

const styles = StyleSheet.create({
  textHeading: {
    fontSize: fontSize.header,
    color: colors.textColor1,
    fontWeight: "bold"
  }
});

HeaderText.propTypes = {
  style: PropTypes.object,
  children: PropTypes.any
}

export default HeaderText;