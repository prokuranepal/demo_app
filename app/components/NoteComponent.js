import React from "react";
import { Text, StyleSheet, View } from "react-native";

import PropTypes from 'prop-types';

import {fontSize} from '../theme/fonts';
import {colors} from '../theme/themeColors';

const NoteComponent = props => (
    <View  style={{...styles.noteView, borderLeftColor:props.color}} data-test="noteComp" >
        <Text style={[styles.textHeading, props.style]} data-test="headingComp">
             {props.title}
         </Text>
        <Text style={[styles.textNormal, props.style]} data-test="descComp">
             {props.children}
         </Text>
  </View>
);

const styles = StyleSheet.create({
    noteView:{
    padding:10,
    paddingTop:12,
    marginTop:10,
    marginBottom:20,
    borderLeftColor: colors.defaultBorder,
    borderLeftWidth:10  ,
    backgroundColor: colors.note
    },

  textNormal: {
    fontSize:fontSize.noteDescription, 
    color:colors.textColor1, 
  },
  
  textHeading: {
    fontSize:fontSize.noteHeading, 
    color: colors.textColor1, 
    fontWeight:"bold"
  }
});

NoteComponent.propTypes = {
  style: PropTypes.object,
  children: PropTypes.any,
  color: PropTypes.string,
  title: PropTypes.string
}

export default NoteComponent;