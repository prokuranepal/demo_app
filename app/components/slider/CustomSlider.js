import React, { useState } from 'react';
import { StyleSheet, View, Text } from 'react-native';

import PropTypes from 'prop-types';

import Slider from './SliderContainer';;

const CustomSlider = props => {

  const maximumValue = props.preferences.length - 1;
  const length = props.preferences.length;
  const sliderRatio = maximumValue / (length - 1);

  const [state, setState] = useState({
    value: props.initialValue,
    adjustSign: 1
  });

  const choice = (v) => {
    const halfRatio = sliderRatio / 2;
    let i = 0;
    for (; ;) {
      if ((v < sliderRatio) || (v <= 0)) {
        if (v >= halfRatio) {
          i++;
        }
        break;
      }
      v = v - sliderRatio;
      i++;
    }
    let value = sliderRatio * i;

    //Move the slider
    value = value + (state.adjustSign * 0.000001);//enforce UI update
    let sign = 1;
    if (state.adjustSign > 0) {
      sign = -1
    } else {
      sign = 1
    }
    // console.log(value, v)
    setState({ ...state, value: value, adjustSign: sign });
    props.selectState(Math.round(value));
  }

  

  return (
    <View style={styles.mainView}>
      <View style={{...styles.sliderContainer,borderColor: props.color}}>

        <Slider
          color={props.color}
          value={state.value}
          minimumValue={0}
          maximumValue={maximumValue}
          onSlidingComplete={(value) => choice(value)}
          trackStyle={{...styles.track,backgroundColor: props.color}}
          thumbStyle={{...styles.thumb,backgroundColor: props.color}}
          minimumTrackTintColor="#464F7A"
          maximumTrackTintColor="#464F7A"
          length={length}
          data-test="sliderComp"
        />

      </View>
      <View style={styles.itemWrapper}>
        {props.preferences.map((option) => {
          return <Text key={option._id} data-test="textComp">{option.sliderText}</Text>
        })}
      </View>
    </View>
  )
};

const styles = StyleSheet.create({
  mainView: {
    flex:1, justifyContent:"center"
  },
  itemWrapper: {
    marginLeft: 8,
    justifyContent: 'space-between',
    alignSelf: 'stretch',
    flexDirection: 'row',
    marginBottom:20,
    marginTop:8
  },
  sliderContainer: {
    marginTop: 20,
    height: 30,
    display: 'flex',
    justifyContent: 'center',
    borderLeftWidth: 15,
    borderRightWidth: 15,
    borderRadius: 16,
  },
  track: {
    height: 30,
    display: 'flex',
    justifyContent: 'center',
  },
  thumb: {
    width: 20,
    height: 20,  
    borderColor: 'white',
    borderWidth: 7,
    borderRadius: 10,
  }
});

CustomSlider.propTypes = {
  preferences: PropTypes.array,
  color: PropTypes.string,
  initialValue: PropTypes.number,
  selectState: PropTypes.func
}


export default CustomSlider;