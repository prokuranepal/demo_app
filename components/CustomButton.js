import React from 'react';
import { View, Button, TouchableOpacity, Text,StyleSheet } from 'react-native';
const customButton =(props)=>(
    <View style={props.container_style}>
        <TouchableOpacity
          style={[styles.saveStyle,props.buttonStyle]}
          onPress={ props.pressHandler }>
            <Text style={{color:props.color, fontSize:17}}> {props.title} </Text>
      </TouchableOpacity>
</View>
)

const styles = StyleSheet.create({
saveStyle:{
    borderRadius:40,
    paddingHorizontal:60,
    paddingVertical:10,
    backgroundColor:"#4fbed2"
  }
})

export default customButton;