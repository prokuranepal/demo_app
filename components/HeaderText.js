import React from "react";
import { Text, StyleSheet } from "react-native";
import {colors} from '../ThemeColors/themeColors'
const headerText = props => (
  <Text {...props} style={[styles.textHeading, props.style]}>
    {props.children}
  </Text>
);

const styles = StyleSheet.create({
  textHeading: {
    fontSize: 26,
    color: colors.textColor1,
    fontWeight: "bold"
  }
});

export default headerText;