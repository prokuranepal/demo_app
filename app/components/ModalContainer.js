import React, { useState } from "react";
import {
    View,
    Modal,
    TouchableOpacity, StyleSheet,
    ScrollView,
    Dimensions
} from "react-native";

import PropTypes from 'prop-types';

import CustomButton from './CustomButton';
import HeaderText from './HeaderText';
import NoteComponent from './NoteComponent';
import CustomSlider from './slider/CustomSlider';
import { useOrientation } from './useOrientation'

import {fontSize} from '../theme/fonts';
import { colors } from '../theme/themeColors';


const ModalContainer = props => {
    const [preferenceOption, setpreferenceOption] = useState(props.diet.preferences[0]);
    
    const orientation = useOrientation();
    const dim = Dimensions.get('screen');
    const dimA = Dimensions.get("window");
    // console.log(dim.height, dim.width, orientation)
    let height = dim.height < 700 ? 152 : 145;
    let buttonFlexSize = dim.height < 700 ? 5 : 3.7;
    let aspectRatio = orientation == "PORTRAIT" ? dimA.height / dimA.width : ((dimA.height / dimA.width) * 3);
    
    const optionSelectHandler = (value) => {
        setpreferenceOption(props.diet.preferences[value]);

    }
    return (

        <Modal
            animationType={"slide"}
            onRequestClose={() => props.actionClick(0, props.diet.dietaryHeader)}
            transparent={true}
            visible={props.isModal}
            data-test="modalComp"
            >

            <TouchableOpacity style={styles.touchable} onPressOut={() => props.actionClick(0, props.diet.dietaryHeader)} data-test="touchableComp"/>
            <View style={styles.modal}>
                <ScrollView contentContainerStyle={styles.scrollView} >

                    <View style={{ height: height * aspectRatio }} data-test="ratioComp">
                        <HeaderText data-test="headerComp">
                            {props.diet.dietaryHeader}
                        </HeaderText>
                        <HeaderText style={styles.description} data-test="descComp">
                            {props.diet.dietaryDescription}
                        </HeaderText>
                        <NoteComponent title={preferenceOption.subHeader} color={preferenceOption.color} data-test="subHeaderDescComp">
                            {preferenceOption.subHeaderDescription}
                        </NoteComponent>
                    </View>

                    <CustomSlider selectState={optionSelectHandler}
                        color={preferenceOption.color}
                        initialValue={preferenceOption._id}
                        preferences={props.diet.preferences}
                        data-test="sliderComp"
                    />

                    <View style={styles.buttonConView}>
                        <CustomButton container_style={{ ...styles.buttonContainer, flex: buttonFlexSize, alignItems: "center" }}
                            buttonStyle={{ backgroundColor: preferenceOption.color }}
                            data-test="saveComp"
                            title="save choice"
                            color={colors.buttonTextColor}
                            pressHandler={() => props.actionClick(preferenceOption._id, props.diet.dietaryHeader)} />
                        <CustomButton container_style={{ ...styles.buttonContainer, flex: 1, alignItems: "flex-start" }}
                            buttonStyle={{ ...styles.buttonCancel, backgroundColor: preferenceOption.color }}
                            data-test="saveComp"
                            title="X"
                            data-test="cancelComp"
                            color={colors.buttonTextColor}
                            pressHandler={() => props.actionClick(0, props.diet.dietaryHeader)} />
                    </View>
                </ScrollView>
            </View>
        </Modal>
    )
};
const styles = StyleSheet.create({

    buttonCancel: { paddingHorizontal: 17, paddingVertical: 10, borderRadius: 40 },
    buttonContainer: {
        flex: 1,
        justifyContent: 'flex-end',
        alignItems: 'center',
    },
    buttonConView: { flex: 0.5, flexDirection: "row", justifyContent: "center", alignItems: "flex-end", marginBottom: 40 },
    description: { fontSize: fontSize.description, paddingVertical: 10, fontWeight: "normal" },
    modal: {
        flex: 3,
        paddingHorizontal: 30,
        paddingTop: 40,
        borderTopRightRadius: 30,
        borderTopLeftRadius: 30,
        backgroundColor: '#f4f4fc',
    },
    scrollView: { flexGrow: 1, paddingHorizontal: 10 },
    touchable: { flex: 1, opacity: 0 },
});

ModalContainer.propTypes = {
    diet: PropTypes.object,
    isModal: PropTypes.bool,
    actionClick: PropTypes.func
}

export default ModalContainer;