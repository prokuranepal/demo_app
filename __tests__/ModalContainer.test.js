import React from 'react';
import {
    configure,
    shallow
} from 'enzyme';
import {View, Modal, ScrollView} from 'react-native';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import HeaderText from '../app/components/HeaderText';
import CustomButton from '../app/components/CustomButton';
jest.mock("Dimensions")

import ModalContainer from '../app/components/ModalContainer';
import NoteComponent from '../app/components/NoteComponent';
import CustomSlider from '../app/components/slider/CustomSlider'
import {act} from 'react-dom/test-utils';
import {findByTestAttr} from '../app/utils/testFunction';
// JestHook.mock('expo-font');
configure({
    adapter: new EnzymeAdapter
})

const function_click=jest.fn()
const dummy_data = {
    actionClick:function_click,
    isModal:false,
    diet:{
        
        dietaryHeader:"Header dietary 1",
        dietaryDescription:"dietary Description 1",
        preferences:[
            {   _id:0,
                subHeader:"preferences 1",
                subHeaderDescription:"sub head description 1",
                color:"red"
            },
            {
                _id:1,
                subHeader:"preferences 2",
                subHeaderDescription:"sub head description 2",
                color:"blue"
            }]
    }
}



describe('<ModalContainer/>', () => {

    afterEach(() => {
        jest.clearAllMocks();
      });

    it('should render without issues', () => {
        const component = shallow(<ModalContainer {...dummy_data}/>);
        expect(component.length).toBe(1);
        expect(component).toMatchSnapshot();
      });

    it("all components test in ModalContainer", () =>
      {
          const wrapper = shallow(<ModalContainer {...dummy_data}>Headerrr</ModalContainer>)
          expect(wrapper.find(HeaderText)).toHaveLength(2)
          expect(wrapper.find(CustomButton)).toHaveLength(2)
          expect(wrapper.find(View)).toHaveLength(3)
          expect(wrapper.find(NoteComponent)).toHaveLength(1)
          expect(wrapper.find(CustomSlider)).toHaveLength(1)
          expect(wrapper.find(ScrollView)).toHaveLength(1)
          expect(wrapper.find(Modal)).toHaveLength(1)
     });
    it("Modal Component test in ModalContainer", () =>
      {
          const wrapper = shallow(<ModalContainer {...dummy_data}>Headerrr</ModalContainer>)
          const modalComp = findByTestAttr(wrapper,"modalComp");
          expect(modalComp).toHaveLength(1)
          expect(modalComp.props().visible).toEqual(false)
          modalComp.props().onRequestClose()
          wrapper.update()
          expect(function_click).toHaveBeenCalledTimes(1)
        //   const wrapper2 = shallow(<ModalContainer {...{...dummy_data, visible:true}}>Headerrr</ModalContainer>)
        //   const modalComp2 = findByTestAttr(wrapper2,"modalComp");
        //   expect(modalComp2.props().visible).toEqual(true)

      
     });

   
     it("Touchable Opacity test in ModalContainer", () =>
     {
         const wrapper = shallow(<ModalContainer {...dummy_data}>Headerrr</ModalContainer>)
         const touchableComp = findByTestAttr(wrapper,"touchableComp");
         expect(touchableComp).toHaveLength(1)
         touchableComp.props().onPressOut()
         wrapper.update()
         expect(function_click).toHaveBeenCalledTimes(1);
     
    });

    it("Header Component test in ModalContainer", () =>
    {
        const wrapper = shallow(<ModalContainer {...dummy_data}>Headerrr</ModalContainer>)
        const headerComp = findByTestAttr(wrapper,"headerComp");
        expect(headerComp).toHaveLength(1)
        expect(headerComp.props().children).toEqual("Header dietary 1")
   });

   it("Header Description Component test in ModalContainer", () =>
   {
       const wrapper = shallow(<ModalContainer {...dummy_data}>Headerrr</ModalContainer>)
       const descComp = findByTestAttr(wrapper,"descComp");
       expect(descComp).toHaveLength(1)
       expect(descComp.props().children).toEqual("dietary Description 1")
  });
  it("NoteComponent and sliderComp test in ModalContainer", () =>
  {
      const wrapper = shallow(<ModalContainer {...dummy_data}>Headerrr</ModalContainer>)
      let subHeaderDescComp = findByTestAttr(wrapper,"subHeaderDescComp");
      expect(subHeaderDescComp).toHaveLength(1)
      expect(subHeaderDescComp.props().title).toEqual("preferences 1")
      expect(subHeaderDescComp.props().children).toEqual("sub head description 1")
      let sliderComp = findByTestAttr(wrapper,"sliderComp");
      expect(sliderComp).toHaveLength(1)
      expect(subHeaderDescComp.props().color).toEqual("red")
      expect(sliderComp.props().initialValue).toEqual(0)
      expect(sliderComp.props().preferences).toEqual([
        {
            _id: 0,
            color: "red",        
            subHeader:"preferences 1",
            subHeaderDescription:"sub head description 1"
        },
        {
            _id: 1,
            color: "blue",        
            subHeader:"preferences 2",
            subHeaderDescription:"sub head description 2"
        }])
    //  const sliderComp2 = findByTestAttr(wrapper,"sliderComp");
     act(() => {sliderComp.props().selectState(1)})
     wrapper.update()
     subHeaderDescComp = findByTestAttr(wrapper,"subHeaderDescComp");
     sliderComp = findByTestAttr(wrapper,"sliderComp");
     expect(subHeaderDescComp.props().color).toEqual("blue")
     expect(sliderComp.props().initialValue).toEqual(1)
     expect(subHeaderDescComp.props().title).toEqual("preferences 2")
     expect(subHeaderDescComp.props().children).toEqual("sub head description 2")
 
});

it("Custom buttons for save and cancel test in ModalContainer", () =>
{
    const wrapper = shallow(<ModalContainer {...dummy_data}>Headerrr</ModalContainer>)
    const saveComp = findByTestAttr(wrapper,"saveComp");
    const cancelComp = findByTestAttr(wrapper,"cancelComp");
    expect(saveComp.props().title).toEqual("save choice")
    expect(cancelComp.props().title).toEqual("X")
    expect(saveComp).toHaveLength(1)
    saveComp.props().pressHandler()
    wrapper.update()
    expect(function_click).toHaveBeenCalledTimes(1);
    cancelComp.props().pressHandler()
    wrapper.update()
    expect(function_click).toHaveBeenCalledTimes(2);

});
// it("Orientation data test in ModalContainer", () =>
// {
//     const wrapper = shallow(<ModalContainer {...dummy_data}>Headerrr</ModalContainer>)
//     // const saveComp = findByTestAttr(wrapper,"saveComp");
//     const ratioComp = findByTestAttr(wrapper,"ratioComp");
//     // expect(saveComp.props().title).toEqual("save choice")
//     // expect(saveComp.props().title).toEqual("X")
//     let mock=jest.fn()
 
//     mock.mockReturnValue("PORTRAIT");
//     // const utils = jest.createMockFromModule('react-native/Libraries/Utilities/Dimensions').default;
//     // utils.get = jest.fn(secret => {return {height:620, width:360}});
//     mockOrientation();
//     // wrapper.instance().useOrientation = mock;
//     mockDimension()
    
//     wrapper.update()
//     expect(ratioComp.props().style).toEqual({height:152*(620/360)})


 
// });


})

   
// export const mockDimension = () => {

//     jest.resetModules();
//     jest.doMock('react-native/Libraries/Utilities/Dimensions', () => ({
//       OS,
//       get: (val) => {
//         console.log("hello testt")  
//         return {height:620, width:360}},
//     }));
//   };

//   export const mockOrientation = () => {
//     jest.resetModules();
//     jest.doMock('../src/components/useOrientation', () => "PORTRAIT"
//     );
//   };