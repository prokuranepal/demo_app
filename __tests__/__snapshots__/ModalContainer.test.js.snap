// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalContainer/> should render without issues 1`] = `ShallowWrapper {}`;
